File: ExpressionAddOperation.java
/*
Desc : We check fo rall combinations of the numbers in the num string and the operators
for every wrong combination we backtrack and proceed the dfs.
Time Complexity : O(2^n)
Space Complexity : O(1)
*/

class Solution {
    List<String> result;
    public List<String> addOperators(String num, int target) {
        result = new ArrayList<>();
        if(num == null || num.length()==0) return result;
        helper(num,target,"",0,0,0);
        return result;
    }
    private void helper(String num, int target,String path,long calc,long tail,int index){
        //base
        if(index == num.length()){
            if(target == calc){
                result.add(path);
                return;
            }
        }
        //logic
        for(int i=index;i<num.length();i++){
            if(num.charAt(index)=='0' && index!=i) continue;
            long curr = Long.parseLong(num.substring(index,i+1));
            if(index == 0){
                helper(num,target,path+curr,curr,curr,i+1);
            }else{
                //+
                helper(num,target,path+'+'+curr,calc+curr,curr,i+1);
                //-
                helper(num,target,path+'-'+curr,calc-curr,-curr,i+1);
                //*
                helper(num,target,path+'*'+curr,calc-tail+tail*curr,tail*curr,i+1);
            }
        }
    }
}

File: expression-add-operation.py
# TC:4^n
# SC:4^n

class Solution:
    def addOperators(self, num: str, target: int) -> List[str]:
        n=len(num)
        output=[]
        
        def dfs(i,so_far_list,res,prev):
            if i>=n:
                #print(so_far_list,res)
                if res==target:
                    output.append("".join(so_far_list))
                    #ouptut.append("".join(so_far_list))
                return
            
            for j in range(i,n):
                cand=int(num[i:j+1])
                if not so_far_list:
                    dfs(j+1,[num[i:j+1]],cand,cand)
                else:
                    dfs(j+1,so_far_list+["+"]+[num[i:j+1]],res+cand,cand)
                    dfs(j+1,so_far_list+["-"]+[num[i:j+1]],res-cand,-cand)
                    dfs(j+1,so_far_list+["*"]+[num[i:j+1]],res-prev+cand*prev,cand*prev)
                if num[i]=="0":
                    break
        
        dfs(0,[],0,None)
        return output